spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/metadata/ZookeeperMetadataStore.java

```diff
@@ -0,0 +1,381 @@
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import com.gs.collections.impl.map.mutable.ConcurrentHashMap;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.ChildData;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheEvent;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
+import org.apache.curator.utils.CloseableUtils;
+import org.apache.curator.utils.EnsurePath;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.util.Assert;
+
+/**
+ * Zookeeper-based {@link ListenableMetadataStore} based on a Zookeeper node. Values are stored in the children node,
+ * the names of which are stored as keys.
+ *
+ * @author Marius Bogoevici
+ */
+public class ZookeeperMetadataStore implements ListenableMetadataStore, SmartLifecycle {
+
+	private final Object lifecycleMonitor = new Object();
+
+	private final CuratorFramework client;
+
+	private final List<MetadataStoreListener> listeners = new CopyOnWriteArrayList<>();
+
+	/**
+	 * An internal map storing local updates, ensuring that they have precedence if the cache contains stale data.
+	 * As changes are propagated back from Zookeeper to the cache, entries are removed.
+	 */
+	private final ConcurrentMap<String, LocalChildData> updateMap = new ConcurrentHashMap<>();
+
+	private volatile String root = "/SpringIntegration-MetadataStore";
+
+	private volatile String encoding = "UTF-8";
+
+	private volatile PathChildrenCache cache;
+
+	private volatile boolean running = false;
+
+	private volatile boolean autoStartup = false;
+
+	private volatile int phase = Integer.MAX_VALUE;
+
+	public ZookeeperMetadataStore(CuratorFramework client) throws Exception {
+		Assert.notNull(client, "Client cannot be null");
+		this.client = client;
+	}
+
+	/**
+	 * Encoding to use when storing data in ZooKeeper
+	 *
+	 * @param encoding encoding as text
+	 */
+	public void setEncoding(String encoding) {
+		Assert.notNull(encoding, "'encoding' must not be null.");
+		this.encoding = encoding;
+	}
+
+	/**
+	 * Root node - store entries are children of this node
+	 *
+	 * @param root encoding as text
+	 */
+	public void setRoot(String root) {
+		Assert.notNull(root, "'root' must not be null.");
+		Assert.isTrue(root.startsWith("/"), "'root' must start with '/'");
+		Assert.isTrue(!root.endsWith("/"), "'root' must not end with '/'");
+		this.root = root;
+	}
+
+	public String getRoot() {
+		return this.root;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public String putIfAbsent(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
```

Comment:
So, is `Assert.hasText` your mental blocker? Or have you really done that intentionally?